PEP 678: more rejected ideas (#2205)
python/peps: pep-0678.rst

========
Exception objects are typically initialized with a message that describes the
error which has occurred.  Because further information may be available when the
exception is caught and re-raised, or included in an ``ExceptionGroup`` , this PEP
proposes to add a ``.__note__`` attribute and update the builtin traceback formatting
code to include it in the formatted traceback following the exception string.

> This is particularly useful in relation to :pep:`654` ``ExceptionGroup`` s, which

Reviewer: I can't test right now, but is this space required for Docutils to process the literal correctly? If it is, could you backslash escape it so that it renders properly?
```suggestion
This is particularly useful in relation to :pep:`654` ``ExceptionGroup``\ s, which
```

Reviewer 2: > I can't test right now, but is this space required for Docutils to process the literal correctly?

Yes, unfortunately—in fact, it results in a hard error when parsing—but that's a good trick to avoid it.

- testing libraries may wish to show the values involved in a failing assertion,
  or the steps to reproduce a failure (e.g. ``pytest`` and ``hypothesis`` ; example below).
- code with retries may wish to note which iteration or timestamp raised which

>   of several possible errors - especially if re-raising them in an ``ExceptionGroup``

Reviewer: This is confusing -- I think the intent is that there is a set of errors that may possibly be raised, and `__note__` forms a way to map loops / timestamps to a specific error raised, but it isn't very clear from the text. Would you be able to rephrase, please?

        +------------------------------------

Non-goals
---------
``__note__`` is *not* intended to carry structured data.  If your note is for use by
a program rather than display to a human, we recommend instead chosing a convention
for an attribute like e.g. ``err._parse_errors = ...`` on the error or ``ExceptionGroup``
(`discussion <https://discuss.python.org/t/accepting-pep-654-exception-groups-and-except/10813/26>`__,

> `discussion <https://bugs.python.org/issue46431>`__).

Reviewer: This might be a case where a footnote is more appropriate? Two links that just say "discussion" are somewhat unhelpful to the reader to tell if I need to care about them / read the entire thread / etc.

Reviewer 2: I second this

As a rule of thumb, prefer `exception chaining <https://docs.python.org/3/tutorial/errors.html#exception-chaining>`__

> when the error is going to be re-raised or handled as an individual error; and prefer

Reviewer: It may be better just to split this into multiple sentences, as there are at least two ideas here.
```suggestion
when the error is going to be re-raised or handled as an individual error, and prefer
```

Reviewer 2: The comma is probably a little more conventional, but the semicolon isn't strictly _wrong_ per say, especially if the writer wants to convey more distance between the two ideas, but not so much as a period. To me it reads fine as one sentence, and splitting it into two would be choppy and repetitive without some rewriting.

> ``__note__`` when you are collecting multiple exception objects to handle together or later.

Reviewer 3: Here you could reference Raymond's post that kicked off the discussion about chaining as relevant history.

        +------------------------------------

Non-goals
---------
``__note__`` is *not* intended to carry structured data.  If your note is for use by

> a program rather than display to a human, we recommend instead chosing a convention

Reviewer 4: ```suggestion
a program rather than display to a human, we recommend instead choosing a convention
```

    Explanation:                                                            # Hence this PEP!
    You can reproduce this error by ...

> **Where these factors are not problematic, we encourage use of exception chaining

Reviewer: What does "problematic" mean here?

Reviewer 2: Its referring to the two factors mentioned in the two paragraphs above the two code blocks, which are admittedly a bit distant (and could justify, for example, something more like

```suggestion
**Where the two factors mentioned above are not problematic, we encourage use of exception chaining
```

to make it more clear.

Reviewer: I wasn't very clear -- my comment was more that "problematic" as a word is quite vague (and not very common) -- I'd suggest something like
```suggestion
**Where these two problems are not relevant, we encourage using `exception chaining`_.**
```
This uses the "problems" phrasing from the lead paragraph, and uses more explicit / plainer English.

A

Reviewer 2: I hear "problematic" used all the time in AmE, so maybe its an AmE-BrE distinction (though I'd think its meaning would seem to be pretty clear, so long as readers are familiar with the root word and the suffix).

Reviewer: Fair enough. Not major really. (This might also be a circles one moves in thing? Anyhow)
A

Reviewer 2: I think its a fair point to make; if it isn't going to be understandable to a large fraction of readers especially internationally, then it should be revised, and assuming we want to do so, your edit is a good way to accomplish that (and also adds the link and clarifies the reference somewhat)

applicable.

Augment the ``raise`` statement
-------------------------------
One discussion proposed ``raise Exception() with "note contents"`` .
This would be a strictly additional feature over adding the attribute and modifying
the traceback-formatting code, and we do not expect ``__note__`` to be so

> frequently-used as to justify new syntax.

Reviewer 3: I think it’s not so much the frequency of use of note, but the way that note will typically be used. Catch-and-reraise-with-note is not the use case that motivated this pep.

Same comment re the decorator idea. Both of these could be added, now or later, but are not necessarily in the scope of this pep.

Reviewer 5: We don't have new syntax for setting the traceback either, but there's a `with_traceback(tb)` method. I could imagine adding a `.with_note(nt)` method, but only if there's a use case that would want to use this rather than assignment to `__note__` (for tracebacks, assignment to `__traceback__` works too, but there's a common idiom where the method is useful).

Reviewer 3: > I think it’s not so much the frequency of use of note, but the way that note will typically be used. Catch-and-reraise-with-note is not the use case that motivated this pep.

On a related point - I'd reword the second sentence of the abstract slightly. It is currently "Because further information may be available when the exception is caught and re-raised...".  Maybe remove the 'and reraised' because that doesn't describe the ExceptionGroups use case.

Reviewer 5: >  I'd reword the second sentence of the abstract slightly. [...]

Makes sense. (Although the reader might wonder why bother storing the note on the exception rather than just passing it along to whatever you pass the exception to. The answer is exactly ExceptionGroup, so maybe mention that explicitly?)

Structured API rather than string attribute
-------------------------------------------
Instead of a string-or-None, ``__note__`` could be a new object with a structured API
and a ``__str__()`` method for printing at trackback time.
We have not identified any scenario where libraries would want to do anything but either
concatenate or replace notes, and so the additional complexity does not seem justified.

Permitting ``__note__`` to be "None or any object with a ``__str__()`` method" would
allow for experimentation with such proposals, and forward-compatibility one was adopted.
We see this as a flexibility/usability tradeoff, and in the absence of any concrete
use-case we narrowly prefer to guarantee that notes are strings but remain open to

> revisiting the decision.

Reviewer 5: This definitely feels like over-engineering. We have no use case but once that feature is added we can never remove it even if it's basically unused.

Add a helper function ``contexlib.add_exc_note()``
--------------------------------------------------
We don't expect notes to be used frequently enough to justify this either, but
provide the following recipe::

    @contextlib.contextmanager
    def add_exc_note(
        note: str,
        exc: type[BaseException] | tuple[type[BaseException], ...] = BaseException,
        concatenate: str | None = "\n\n",
    ):
        try:
            yield
        except exc as err:
            if concatenate is None or err.__note__ is None:
                err.__note__ = note
            else:
                err.__note__ = err.__note__ + concatenate + note

>             raise

Reviewer 5: Maybe you can provide a simple example of how this would be used? I'd also trim down the recipe down significantly -- people who want to constrain the exception or tweak the concatenation behavior can extend the recipe.

applicable.

Store notes in ``ExceptionGroup``\ s
------------------------------------
Initial discussions proposed making a more focussed change by thinking about how to
associate messages with the nested exceptions in ``ExceptionGroup`` s, such as a list
of notes or mapping of exceptions to notes.  However, this would force a remarkably
awkward API and retains a lesser form of the cross-referencing problem discussed
under "use ``print()``" above; if this PEP is rejected we prefer the status quo.
Finally, of course, ``__note__`` is not only useful with ``ExceptionGroup`` s!

Possible Future Enhancements
============================

> In addition to rejected alternatives, there have been a range of suggestions which

Reviewer 3: ```suggestion
In addition to rejected alternatives, there have been a range of enhancements which
```

Maye this actually

> we suggest should be deferred to a future version, when we have more experience with

Reviewer 3: ```suggestion
we believe should be deferred to a future version, when we have more experience with
```

(just so you don't use suggestions and suggest in the same sentence)

the uses (and perhaps misuses) of ``__note__``.

Allow any object, and cast to string for display
------------------------------------------------
We have not identified any scenario where libraries would want to do anything but either
concatenate or replace notes, and so the additional complexity and interoperability
challenges do not seem justified.

Permitting any object would also force any future structured API to change the behaviour
of already-legal code, whereas expanding the legal contents of ``__note__`` from strings
to include other objects is fully backwards-compatible.  In the absence of any proposed

> use-case (see also `Non-goals`_), we strongly prefer to keep things simple.

Reviewer 3: ```suggestion
use-case (see also `Non-goals`_), we prefer to begin with a restrictive API that can be relaxed later.
```

under "use ``print()``" above; if this PEP is rejected we prefer the status quo.
Finally, of course, ``__note__`` is not only useful with ``ExceptionGroup`` s!

Possible Future Enhancements
============================

In addition to rejected alternatives, there have been a range of suggestions which
we believe should be deferred to a future version, when we have more experience with
the uses (and perhaps misuses) of ``__note__``.

Allow any object, and cast to string for display
------------------------------------------------
We have not identified any scenario where libraries would want to do anything but either
concatenate or replace notes, and so the additional complexity and interoperability
challenges do not seem justified.

Permitting any object would also force any future structured API to change the behaviour
of already-legal code, whereas expanding the permitted contents of ``__note__`` from strings
to include other objects is fully backwards-compatible.  In the absence of any proposed
use-case (see also `Non-goals`_), we prefer to begin with a restrictive API that can
be relaxed later.

> Add a helper function ``contexlib.add_exc_note()``
> --------------------------------------------------

Reviewer 4: ```suggestion
Add a helper function ``contextlib.add_exc_note()``
---------------------------------------------------
```

Initial discussions proposed making a more focussed change by thinking about how to
associate messages with the nested exceptions in ``ExceptionGroup`` s, such as a list
of notes or mapping of exceptions to notes.  However, this would force a remarkably
awkward API and retains a lesser form of the cross-referencing problem discussed
under "use ``print()``" above; if this PEP is rejected we prefer the status quo.
Finally, of course, ``__note__`` is not only useful with ``ExceptionGroup`` s!

Possible Future Enhancements
============================

In addition to rejected alternatives, there have been a range of suggestions which
we believe should be deferred to a future version, when we have more experience with
the uses (and perhaps misuses) of ``__note__``.

Allow any object, and cast to string for display
------------------------------------------------
We have not identified any scenario where libraries would want to do anything but either
concatenate or replace notes, and so the additional complexity and interoperability
challenges do not seem justified.

Permitting any object would also force any future structured API to change the behaviour

> of already-legal code, whereas expanding the legal contents of ``__note__`` from strings

Reviewer 3: ```suggestion
of already-legal code, whereas expanding the permitted contents of ``__note__`` from strings
```

to include other objects is fully backwards-compatible.  In the absence of any proposed
use-case (see also `Non-goals`_), we strongly prefer to keep things simple.

Add a helper function ``contexlib.add_exc_note()``
--------------------------------------------------

> We don't expect notes to be used frequently enough to justify this either, but

Reviewer 3: ```suggestion
We don't expect notes to be used frequently enough to justify this either, but
```

We don't know how frequently it will be used. How about this:

It was suggested that we add a utility such as the one below to contextlib. We are open to this suggestion, but we believe this is not a core part of the proposal of this PEP and can be added later as an enhancement.

Reviewer 3: Thanks for the thumbs-up. As discussed by email, I will commit the remaining few tweaks. Good night.

> We don't expect notes to be used frequently enough to justify this either, but
> provide the following extensible recipe:

Reviewer 3: ```suggestion
It was suggested that we add a utility such as the one below to the standard
library. We are open to this idea, but do not see it as a core part of the
proposal of this PEP as it can be added as an enhancement later.
```

.. code-block:: python

    @contextlib.contextmanager
    def add_exc_note(note: str):
        try:
            yield
        except Exception as err:
            if err.__note__ is None:
                err.__note__ = note
            else:
                err.__note__ = err.__note__ + "\n\n" + note
            raise

    with add_exc_note(f"While attempting to frobnicate {item=}"):
        frobnicate_or_raise(item)

Augment the ``raise`` statement
-------------------------------
One discussion proposed ``raise Exception() with "note contents"``, but this
does not address the original motivation of compatibility with ``ExceptionGroup``.

> Furthermore, this would be an additional feature over adding the attribute and
> modifying the traceback-formatting code, and we do not expect ``__note__`` to be
> so frequently-used as to justify new syntax, or even a new ``.with_note()`` method
> analogous to ``.with_traceback()``.

Reviewer 3: ```suggestion
Furthermore, we do not believe that the problem we are solving requires or justifies
new language syntax.
```

> 
> 

Reviewer 3: ```suggestion

```

applicable.

Augment the ``raise`` statement
-------------------------------
One discussion proposed ``raise Exception() with "note contents"`` , but this
does not address the original motivation of compatibility with ``ExceptionGroup`` .

Furthermore, this would be an additional feature over adding the attribute and
modifying the traceback-formatting code, and we do not expect ``__note__`` to be

> so frequently-used as to justify new syntax or even a new ``.with_note()`` method

Reviewer 2: ```suggestion
so frequently-used as to justify new syntax, or even a new ``.with_note()`` method
```

Could use a comma here

> (analogous to ``.with_traceback()`` ).

Reviewer 2: ```suggestion
(analogous to ``.with_traceback()``).
```

Symbols are fine after ` `` `, just not alphanumeric characters

Allow any object, and cast to string for display
------------------------------------------------
We have not identified any scenario where libraries would want to do anything but either
concatenate or replace notes, and so the additional complexity and interoperability
challenges do not seem justified.

Variations on this have also been proposed for forward-compatibility with a future

> structured API, but in the absence of any proposed use-case (see also "Non-goals")

Reviewer 2: ```suggestion
structured API, but in the absence of any proposed use-case (see also `Non-goals`_)
```

Actually link the section here

we strongly prefer to keep things simple - and note that implementing any nontrival

> future behavior would then break backwards-compatibility.

Reviewer 3: This can also be in a "possible enhancements" section (come up with a use case, then we add it).

> future behavior would then break backwards-compatibility.

Reviewer 2: Besides awkwardness in the syntax, its unclear to me whether this refers to what would happen if you retain the current proposal of only allowing `str` rather than casting to it (which if so, would seem to be somewhat redundant with the first phrase), or is saying that any non-trivial future behavior implemented in a future structured API would break backward compat anyway.

Add a helper function ``contexlib.add_exc_note()``
--------------------------------------------------
We don't expect notes to be used frequently enough to justify this either, but

> provide the following extensible recipe::

Reviewer 3: This is the second time you’re saying you rejected an idea because you don’t think notes will be used frequently enough to justify it. I don’t think this is the reason.

The context manager is an enhancement we could add any time in the future. I don’t think you’re really rejecting it, just not including it in the basic spec of the feature. Maybe you need a section for ‘possible future enhancements’.

> provide the following extensible recipe::

Reviewer 2: ```suggestion
provide the following extensible recipe:

.. code-block:: python
```

To get proper syntax highlighting and match the others

    @contextlib.contextmanager
    def add_exc_note(note: str):
        try:
            yield
        except Exception as err:
            if err.__note__ is None:
                err.__note__ = note
            else:
                err.__note__ = err.__note__ + "\n\n" + note
            raise

    with add_exc_note(f"While attempting to frobnicate {item=}"):
        frobnicate_or_raise(item)

> Store notes in ``ExceptionGroup`` s
> -----------------------------------

Reviewer: Similar to above, if the space is required please could you escape it

```suggestion
Store notes in ``ExceptionGroup``\ s
------------------------------------
```
